contentctl/objects/config.py

```diff
@@ -1,110 +1,106 @@
-    email: Union[ConfigEmail,None] = None
-    slack: Union[ConfigSlack,None] = None
-    phantom: Union[ConfigPhantom,None] = None
-    rba: Union[ConfigRba,None] = None
-
-
-class ConfigAlertAction(BaseModel):
-    notable: ConfigNotable
-
-
-
-
-class ConfigDeploy(BaseModel):
-    description: str = "Description for this deployment target"
-    server: str = "127.0.0.1"
-
-CREDENTIAL_MISSING = "PROVIDE_CREDENTIALS_VIA_CMD_LINE_ARGUMENT"
-class ConfigDeployACS(ConfigDeploy):
-    token: str = CREDENTIAL_MISSING
+from enum import StrEnum, auto
+import pathlib
+from contentctl.helper.utils import Utils
+from urllib.parse import urlparse
+from abc import ABC, abstractmethod
+from contentctl.objects.enums import PostTestBehavior
+from contentctl.input.yml_reader import YmlReader
+from contentctl.objects.detection import Detection
+
+
+ENTERPRISE_SECURITY_UID = 263
+COMMON_INFORMATION_MODEL_UID = 1621
+
+SPLUNKBASE_URL = "https://splunkbase.splunk.com/app/{uid}/release/{version}/download"
+
+class App_Base(BaseModel,ABC):
+    model_config = ConfigDict(use_enum_values=True,validate_default=True, arbitrary_types_allowed=True)
+    uid: Optional[int] = Field(default=None)
+    title: str = Field(description="Human-readable name used by the app. This can have special characters.")
+    appid: Optional[Annotated[str, Field(pattern="^[a-zA-Z0-9_-]+$")]]= Field(default=None,description="Internal name used by your app. "
+                                                                    "It may ONLY have characters, numbers, and underscores. No other characters are allowed.")
+    version: str = Field(description="The version of your Content Pack.  This must follow semantic versioning guidelines.")
+    description: Optional[str] = Field(default="description of app",description="Free text description of the Content Pack.")
     
 
-class ConfigDeployRestAPI(ConfigDeploy):
-    port: int = 8089
-    username: str = "admin"
-    password: str = PASSWORD
     
 
-class Deployments(BaseModel):
-    acs_deployments: list[ConfigDeployACS] = []
-    rest_api_deployments: list[ConfigDeployRestAPI] = [ConfigDeployRestAPI()]
+    def getSplunkbasePath(self)->HttpUrl:
+        return HttpUrl(SPLUNKBASE_URL.format(uid=self.uid, release=self.version))
 
+    @abstractmethod
+    def getApp(self, config:test, stage_file:bool=False)->str:
+        ...
 
+    def ensureAppPathExists(self, config:test, stage_file:bool=False):
+        if stage_file:
+            if not config.getLocalAppDir().exists():
+                config.getLocalAppDir().mkdir(parents=True)
 
-class ConfigBuildSplunk(BaseModel):
-    pass
+class TestApp(App_Base):
+    model_config = ConfigDict(use_enum_values=True,validate_default=True, arbitrary_types_allowed=True)
+    hardcoded_path: Optional[Union[FilePath,HttpUrl]] = Field(default=None, description="This may be a relative or absolute link to a file OR an HTTP URL linking to your app.")
```

Comment:
What's the use case for making `hardcodedpath` `Optional`? In what circumstances would you want to define an app w/o a URL or fielpath?